### PR TITLE
Lasting proposals & marking proposals in nexus

### PR DESCRIPTION
**Pull request by Devorein**

### CODE REVIEW

File: background/tasks/sendNotifications/sendNotifications.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ export async function getNotifications (): Promise<PendingTasksProps[]> {
           in: [
             ...gnosisSafeTasks.map(getGnosisSafeTaskId),
             ...voteTasks.map(voteTask => voteTask.id),
-            ...workspaceEvents.map(workspaceEvent => workspaceEvent.id)
+            ...workspaceEvents.map(workspaceEvent => `${workspaceEvent.id}.${user.id}`)
```
Comment (mattcasey) [inline background/tasks/sendNotifications/sendNotifications.ts:75]: do we have the same issue with votes?

File: components/nexus/ProposalTasksList.tsx [hunk 2]
```diff
@@ -110,11 +114,36 @@ export function ProposalTasksListRow (
...
+              unmarked: []
+            }
+          } : undefined;
+        }, {
+          revalidate: false
+        });
+      }
+    }
+    main();
+  }, [tasks, user]);
```
Comment (mattcasey) [inline components/nexus/ProposalTasksList.tsx:146]: we only want to run this if user.id changes

File: lib/proposal/getProposalTasks.ts [hunk 3]
```diff
@@ -27,145 +30,142 @@ export function extractProposalData (proposal: Proposal & {
...
+        id: workspaceEvent?.id ?? proposal.id,
+        pagePath: (proposal.page as Page).path,
+        pageTitle: (proposal.page as Page).title,
+        spaceDomain: proposal.space.domain,
+        spaceName: proposal.space.name,
+        status: proposal.status,
+        action
+      };
+
+      if (workspaceEvent && !userNotificationIds.has(`${workspaceEvent.id}.${userId}`)) {
```
Comment (mattcasey) [inline lib/proposal/getProposalTasks.ts:158]: as mentioned above... the formatting of the id should really only be handled in one place, or in a utility function. that would allow us to see where it gets used at least, and we make sure it's all consistent

File: lib/proposal/getProposalTasksFromWorkspaceEvents.ts [hunk 4]
```diff
@@ -120,7 +120,7 @@ export async function getProposalTasksFromWorkspaceEvents (userId: string, works
       }));
 
       const proposalTask: Omit<ProposalTask, 'action'> = {
-        id: workspaceEvent.id,
+        id: `${workspaceEvent.id}.${userId}`,
```
Comment (mattcasey) [inline lib/proposal/getProposalTasksFromWorkspaceEvents.ts:123]: +1 can we do this inside getProposalTasks as well?

**Review by mattcasey — CHANGES_REQUESTED** [review]
TL;DR: can you reduce the amount of places we format the unique id for proposal tasks. But Also, isn't this an issue for votes? And did you consider just fixing the constraint on the db?

File: background/tasks/sendNotifications/sendNotifications.ts (hunk 1 above)
Comment (Devorein) [reply background/tasks/sendNotifications/sendNotifications.ts:75]: I've removed the userId when creating the workspace event id, it turns out we had two unique directives for our UserNotification prisma schema. Removing that fixed the issue.

File: lib/proposal/getProposalTasks.ts [hunk 5]
```diff
@@ -27,145 +30,142 @@ export function extractProposalData (proposal: Proposal & {
...
+      const action = getProposalAction(
+        {
+          currentStatus: proposal.status,
+          isAuthor,
+          isReviewer
+        }
+      );
+
+      const proposalTask = {
+        id: workspaceEvent?.id ?? proposal.id,
```
[the remark this answers is not in the document]
Comment (Devorein) [reply lib/proposal/getProposalTasks.ts:149]: Updated code to use a uuid, rather than relying on proposal if when workspace event doesn't exist. Normally workspace event does exist (unless its a newly created proposal), but now we will create a new workspace event anytime a proposal is created.

File: lib/proposal/getProposalTasksFromWorkspaceEvents.ts (hunk 4 above)
Comment (Devorein) [reply lib/proposal/getProposalTasksFromWorkspaceEvents.ts]: Removed userId when generating proposalTask id

File: prisma/schema.prisma [hunk 6]
```diff
@@ -139,6 +139,7 @@ enum ProposalStatus {
 
 enum WorkspaceEventType {
   proposal_status_change
+  proposal_create
```
[the remark this answers is not in the document]
Comment (Devorein) [reply prisma/schema.prisma]: Yep I will remove the migration and update the code dependant upon it.

**Review by mattcasey — APPROVED** [review]
